gto/git_utils.py

```diff
@@ -82,140 +92,72 @@ def _call_commit_push(
         tag_name,
-        repo.working_dir,
+        scm.root_dir,
         remote_name,
-        remote.url,
     )
-    remote_push_args = [tag_name]
-    if delete:
-        remote_push_args = ["--delete"] + remote_push_args
-    push_info = remote.push(remote_push_args)
-    if push_info.error is not None:
+    try:
+        result = scm.push_refspecs(remote_name, refspec)
+    except SCMError as e:
         raise GTOException(
-            msg=f"The command `git push {remote_name} {' '.join(remote_push_args)}` failed. "
+            msg=f"The command `git push {remote_name} {tag_name}` failed. "
             f"Make sure your local repository is in sync with the remote."
-        )
-
-
-def git_push(repo: Union[str, git.Repo]) -> None:
-    read_repo(repo).git.push()
+        ) from e
+    for _ref, status in result:
+        if status == SyncStatus.DIVERGED:
+            raise GTOException(
+                msg=f"The command `git push {remote_name} {tag_name}` failed. "
+                f"Make sure your local repository is in sync with the remote."
+            )
 
 
-def git_add_and_commit_all_changes(repo: Union[str, git.Repo], message: str) -> None:
-    repo = read_repo(repo)
-    tracked, untracked = _get_repo_changed_tracked_and_untracked_files(repo=repo)
-    if len(tracked) + len(untracked) > 0:
+def git_add_and_commit_all_changes(scm: Git, message: str) -> None:
+    staged, unstaged, untracked = scm.status()
+    if staged or unstaged or untracked:
+        files = list(chain(staged, unstaged, untracked))
         logging.debug("Adding to the index the untracked files %s", untracked)
-        logging.debug("Add and commit changes to files %s", tracked + untracked)
-        repo.index.add(items=tracked + untracked)
-        repo.index.commit(message=message)
+        logging.debug("Add and commit changes to files %s", files)
+        scm.add(files)
+        scm.commit(message)
 
 
-def read_repo(repo: Union[str, git.Repo], search_parent_directories=False) -> git.Repo:
-    if isinstance(repo, Repo):
-        return repo
-    try:
-        return git.Repo(repo, search_parent_directories=search_parent_directories)
-    except (InvalidGitRepositoryError, NoSuchPathError) as e:
-        raise NoRepo(repo) from e
+def _reset_repo_to_head(scm: Git) -> None:
+    if scm.stash.push(include_untracked=True):
+        scm.stash.drop()
 
 
-@contextmanager
-def stashed_changes(repo: Union[str, git.Repo], include_untracked: bool = False):
-    repo = read_repo(repo)
-    if len(repo.refs) == 0:
-        raise RuntimeError(
-            "Cannot stash because repository has no ref. Please create a first commit."
-        )
-
-    tracked, untracked = _get_repo_changed_tracked_and_untracked_files(repo=repo)
-
-    stash_arguments = ["push"]
-    if include_untracked:
-        stash_arguments += ["--include-untracked"]
-    else:
-        untracked = []
-
+def has_remote(scm: Git, remote: str = "origin") -> bool:
```

Comment:
```suggestion
def is_remote(scm: Git, remote: str = "origin") -> bool:
```
...right?

Reply:
ping @aguschin  - maybe you have some thoughts on the original intent so we can followup if needed